File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamworkshop/callback/UserActionPublishedFilesCallback.java
package in.dragonbra.javasteam.steam.handlers.steamworkshop.callback;

import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientUCMEnumeratePublishedFilesByUserActionResponse;
import in.dragonbra.javasteam.steam.handlers.steamworkshop.EnumerationUserDetails;
import in.dragonbra.javasteam.steam.handlers.steamworkshop.SteamWorkshop;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.JobID;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

/**
 * This callback is received in response to calling {@link SteamWorkshop#enumeratePublishedFilesByUserAction(EnumerationUserDetails)}.
 */
public class UserActionPublishedFilesCallback extends CallbackMsg {

    private EResult result;

    private List<File> files;

    private int totalResults;

    public UserActionPublishedFilesCallback(JobID jobID, CMsgClientUCMEnumeratePublishedFilesByUserActionResponse.Builder msg) {
        setJobID(jobID);

        result = EResult.from(msg.getEresult());

        List<File> fileList = new ArrayList<>();
        for (CMsgClientUCMEnumeratePublishedFilesByUserActionResponse.PublishedFileId f : msg.getPublishedFilesList()) {
            fileList.add(new File(f));
        }
        files = Collections.unmodifiableList(fileList);

        totalResults = msg.getTotalResults();
    }

    /**
     * @return Gets the result.
     */
    public EResult getResult() {
        return result;
    }

    /**
     * @return Gets the list of enumerated files.
     */
    public List<File> getFiles() {
        return files;
    }

    /**
     * @return Gets the count of total results.
     */
    public int getTotalResults() {
        return totalResults;
    }

    /**
     * Represents the details of a single published file.
     */
    public static class File {

        private long fileID;

        private Date timestamp;

        public File(CMsgClientUCMEnumeratePublishedFilesByUserActionResponse.PublishedFileId file) {
            fileID = file.getPublishedFileId();
            timestamp = new Date(file.getRtimeTimeStamp() * 1000L);
        }

        /**
         * @return Gets the file ID.
         */
        public long getFileID() {
            return fileID;
        }

        /**
         * @return Gets the timestamp of this file.
         */
        public Date getTimestamp() {
            return timestamp;
        }
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamworkshop/callback/UserActionPublishedFilesCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamworkshop.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverUcm.CMsgClientUCMEnumeratePublishedFilesByUserActionResponse
import `in`.dragonbra.javasteam.steam.handlers.steamworkshop.SteamWorkshop
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import java.util.*

/**
 * This callback is received in response to calling [SteamWorkshop.enumeratePublishedFilesByUserAction].
 */
@Suppress("MemberVisibilityCanBePrivate", "unused")
class UserActionPublishedFilesCallback(packetMsg: IPacketMsg?) : CallbackMsg() {

    /**
     * Gets the result.
     */
    val result: EResult

    /**
     * Gets the list of enumerated files.
     */
    val files: List<File>

    /**
     * Gets the count of total results.
     */
    val totalResults: Int

    init {
        val response = ClientMsgProtobuf<CMsgClientUCMEnumeratePublishedFilesByUserActionResponse.Builder>(
            CMsgClientUCMEnumeratePublishedFilesByUserActionResponse::class.java,
            packetMsg
        )
        val msg = response.body

        jobID = response.targetJobID

        result = EResult.from(msg.eresult)

        files = msg.publishedFilesList.map { File(it) }

        totalResults = msg.totalResults
    }

    /**
     * Represents the details of a single published file.
     */
    class File(file: CMsgClientUCMEnumeratePublishedFilesByUserActionResponse.PublishedFileId) {

        /**
         * Gets the file ID.
         */
        val fileID: Long = file.publishedFileId

        /**
         * Gets the timestamp of this file.
         */
        val timestamp: Date = Date(file.rtimeTimeStamp * 1000L)
    }
}
